### PR TITLE
[ADVAPI32] Improve handling of unaligned key name in RegOpenKeyExW

### PR DESCRIPTION
**Pull request by tkreuzer**
Check for unaligned buffer before calling NtOpenKey instead of checking the result for STATUS_DATATYPE_MISALIGNMENT.

Test x86: https://reactos.org/testman/compare.php?ids=91100,91107,91111,91113

### CODE REVIEW

**Comment by HBelusca** [comment]
What is the advantage of this solution against the other one? Is it for "optimization" purposes (not calling the whole NtOpenKey stuff yet again)?

**Comment by tkreuzer** [comment]
> What is the advantage of this solution against the other one? Is it for "optimization" purposes (not calling the whole NtOpenKey stuff yet again)?

That too, but mostly to avoid exceptions, when they are not needed, because those make debugging to catch an actual exception ("!gflag +soe") only more difficult.

**Review by JoachimHenze — APPROVED** [review]

File: dll/win32/advapi32/reg/reg.c [hunk 1]
```diff
@@ -3400,8 +3401,27 @@ RegOpenKeyExW(HKEY hKey,
     if (lpSubKey == NULL || wcscmp(lpSubKey, L"\\") == 0)
         RtlInitUnicodeString(&SubKeyString, L"");
     else
+    {
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply dll/win32/advapi32/reg/reg.c:3406]: Fixed.

[the remark this answers is not in the document]
Comment (tkreuzer) [reply dll/win32/advapi32/reg/reg.c]: `RtlDuplicateUnicodeString` uses `SubKeyString`, so this wouldn't help.

**Pull request merged**